### ISSUE TITLE
unable to get local issuer certificate

### ISSUE THREAD

**Issue by mits-techreceptives** [mailgun/mailgun-php#174]
I am getting below error : 
GuzzleHttp\Exception\RequestException
cURL error 60: SSL certificate problem: unable to get local issuer certificate
↵
Caused by: GuzzleHttp\Ring\Exception\RingException

how to fix this ? what parameters i am missing during configuration ?
i set up as shown in doc but still it throws error.
i am on window localhost.

**Comment by DavidGarciaCat** [comment]
IMHO if the error is caused by `GuzzleHttp\Ring\Exception\RingException` you should open an issue on Guzzle HTTP Client repository https://github.com/guzzle/guzzle given Mailgun is not the owner of that project. Regards,

**Comment by mits-techreceptives** [comment]
but the actual thing is , there is no any additional config setting for to set SSL => false .
it's directly handle by mailgun.php class file. so it causes this issue.

**Comment by DavidGarciaCat** [comment]
Could you copy/paste the code that you're using to manage this request please?
```
// your code here
```
so maybe we can check if this is really a Mailgun issue. Thanks,

**Comment by mits-techreceptives** [comment]
I am working on Yii2 based Web application.
below it's my configuration web.php file's setting:

'mailer' => [
			'class' => 'boundstate\mailgun\Mailer',
			'key' => 'key-valid-key-generatedfrom-rackspace-formailgun',
			'domain' => 'sandbox{valid-domain}.mailgun.org',
		],

below is my code to send mail:
Yii::$app->mailer->compose()
    ->setFrom('{valid-email}')
    ->setTo('{valid-email}')
    ->setSubject("test mail from mailgun")
    ->send();

## If i set $ssl=false manually in /mailgun-php/src/Mailgun/Mailgun.php file In
public function __construct($apiKey = null, $apiEndpoint = "api.mailgun.net", $apiVersion = "v2", $ssl = false ){
        $this->apiKey = $apiKey;
        $this->restClient = new RestClient($apiKey, $apiEndpoint, $apiVersion, $ssl);
    }

than only it works. but ideally it this should be manageable from front end configuration as user should not be disturb actual library code.

**Comment by mits-techreceptives** [comment]
does mailgun-php api support non-SSL connections server-side! ??

**Comment by DavidGarciaCat** [comment]
So, if I understand your case, you're using an external repository to manage mail requests (a Framework integration) managed by another user instead of Mailgun (and community), you get an error message from an external package (GuzzleHttp instead of Mailgun) and you're expecting to get support from Mailgun (that is not the owner of the Framework integration that you're using or the HTTP Client that is throwing the error), right?

IMHO your current settings look to complex - I think the easy way to fix your case is to simplify your code, using just the Mailgun-PHP package and then trying to send the email. If you are still getting an error then maybe someone can have a look (providing an example of the used code).

An example: as you can imagine, I cannot go to Microsoft and tell them that my Debian Linux is not able to detect my network card, because Microsoft is not the owner of Debian.

@Nyholm are you agree with me?

**Comment by mits-techreceptives** [comment]
Yes agree....
but i am on my localhost on window machine. and with sendgrid / SES all are working with same way is perfect . so got surprised why this one not working with localhost without https ?

**Comment by narendravaghela** [comment]
If you want, you can install local ssl certificate. See below steps.

1. Open http://curl.haxx.se/ca/cacert.pem
2. Copy the entire page and save it as a "cacert.pem"
3. Open your php.ini file and insert or update the following line: curl.cainfo = "[pathtofile]\cacert.pem"

**Comment by Nyholm** [comment]
I could agree that this issue is unrelated to the mailgun library. @mits-techreceptives Please use the latest version of the library that would give you more control over your HTTP client. 

I've created https://github.com/mailgun/mailgun-php/pull/179 to note that `setSslEnabled` should not be used. 

If you have issues on your localhost with verify SSL. You could follow the advice given in this thread or configure your HTTP client not to verify the SSL protocol. Here is an example with Guzzle. 

```php
$client = new \GuzzleHttp\Client([
    'verify' => false,
]);
$adapter = new \Http\Adapter\Guzzle6\Client($client);
$mailgun = new \Mailgun\Mailgun($mailgunApiKey, $adapter);
```

**Comment by lordspace** [comment]
Is there a workaround? because checking Mailgun class there are deprecated things and will be removed.

```
public function __construct(
        $apiKey = null, /* Deprecated, will be removed in 3.0 */
        HttpClient $httpClient = null,
        $apiEndpoint = 'api.mailgun.net', /* Deprecated, will be removed in 3.0 */
        Hydrator $hydrator = null,
        RequestBuilder $requestBuilder = null
```

**Comment by berksites** [comment]
This issue just popped up for me out of no where - was working fine till yesterday.

**Comment by chuntwalker** [comment]
Same for me. This literally stopped working for me this morning.

**Comment by berksites** [comment]
I think that it has something to do with their cert - I'm waiting for a response -> I hate having to build api redundancy

**Comment by chuntwalker** [comment]
agreed

**Comment by dalton478** [comment]
Same for me. No changes to site and been working for years. Broke since yesterday.

**Comment by berksites** [comment]
got a response from MG:

Hello,

Thank you for reaching out to us in regards to our SSL certification renewal, which are now fully completed. This update was pushed out to correct numerous certificates that did not comply with the industry-developed CA/Browser Forum Baseline Requirements by Symantec’s PKI business. You can read more about this here - https://security.googleblog.com/2017/09/chromes-plan-to-distrust-symantec.html.

Now that the updates have been completed, if you’re still having difficulties sending, we can provide some suggestions to get your application back up and running.

If you’re utilizing PHP-
1) Download the latest CA certificates from (https://curl.haxx.se/docs/caextract.html) This is an extract of a list of certificates vetted by Mozilla.
2) Set the path to the CA certificates in the .htaccess file with a directive like the following:
php_value curl.cainfo "/path/to/cacert.pem"
If you have access to your php.ini file, you can modify this value globally by setting curl.cainfo = "/path/to/cacert.pem"

If you’re utilizing cURL-
Ubuntu users: https://askubuntu.com/questions/646594/how-to-fix-ca-cert-issues-with-curl-in-ubuntu-14-04
RedHat users: https://serverfault.com/questions/394815/how-to-update-curl-ca-bundle-on-redhat
Be sure to restart your sending application after updating.

If a restart of your application does not clear the issue, the certificate cache may need to be cleared. Here is some additional information as to how this might be accomplished for you:

    http://unitstep.net/blog/2009/05/05/using-curl-in-php-to-access-https-ssltls-protected-sites/
    http://stackoverflow.com/questions/7219989/java-and-ssl-certificates


Because every sending application is be different, the exact method for updating the certifications/restarting your sending application may vary for you. If none of these options worked, some independent research might be needed for the update/restart method applicable to you.

While these suggestions are meant to help resolve the issues you are experiencing, making certain changes to your environment can always cause further issues with your application. We highly recommend making these changes in a non-production environment prior to releasing into production. These are only intended as suggested fixes and in no-way guarantee that this will fix the issue, further steps may need to be taken depending on your specific environment. As always Mailgun will always do our best to answer any questions that you may have, but there are limitations in the support we can provide.

Please let us know if you have any questions.

Thanks,
Chris@MG


_________________________________________________________________________________

Now I attempted to do as he instructed but using that cert seems to break my integration with google recaptcha - I think its gonna be a long night - probably end with me ditching mailgun for a bit - we'll see.

**Comment by dalton478** [comment]
I dropped the mailgun-php package and just used curl commands to get around it. Sourced from https://gist.github.com/jonaslejon/e1599d28121294e94601

Replace MAILGUN_API with your key and other obvious mods to get it running quickly.

Also set 
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
so it would ignore the cert issue which is ok in my use case.

**Comment by snowbreeze** [comment]
Same issue, I tried 
```
public function __construct($apiKey = null, $apiEndpoint = "api.mailgun.net", $apiVersion = "v2", $ssl = false){
```
But didnt help. What can we do? I am running on a shared host on php, dont really have access to phpini and all that.

**Comment by DavidGarciaCat** [comment]
Hi all,

Based on the MG response, looks like it confirm what we were discussing on this other issue https://github.com/mailgun/mailgun-php/issues/437 where @izazueta posted this link:

https://twitter.com/joshcarcione/status/955926129893888000

@snowbreeze may I ask what is the version that you’re using? That constructor looks old, we now use a new named constructor to deal with the requests.

Cheers,

**Comment by snowbreeze** [comment]
I dont really want to upgrade the lib anymore, just moved to curl and it fixed my problems.

**Comment by xPrzybyLx** [comment]
So what can I do to make this work for SSL certificate generated for localhost - domain pointed to my localhost ?

**Comment by chuntwalker** [comment]
@berksites thanks for posting this. I was still waiting on my MG support request. This solution worked for me.

**Comment by berksites** [comment]
I don't understand - why do we need to download a cert from a website of self proclaimed hackers?  read http://haxx.se 

I am more leaning towards the whole "clear my cert cache" thing MG mentioned but I am not sure exactly how to go about that.

**Comment by cjcastellanosing** [comment]
Thanks @DavidGarciaCat

In twitter post: "it may be worth mentioning that users who are using an old API package containing guzzle will need to update the cacert file in the resources directory. u can download it from https://curl.haxx.se/ca/cacert.pem" 

It fixed my problems.

**Comment by dwyer14** [comment]
+1 for @cjcastellanosing's solution/note. Updating the `cacert.pem` file at `/vendor/guzzle/guzzle/src/Guzzle/Http/Resources` did the trick for me. Thanks for sharing that!

**Comment by obcidio** [comment]
Issue popped out of no where for me. @dalton478's suggestion to set verify_peer worked for me.
`curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);`

**Comment by michaelphipps** [comment]
@berksites 
> I don't understand - why do we need to download a cert from a website of self proclaimed hackers? read http://haxx.se

Just FYI: Hacker != cracker.  The guys at haxx.se don't have malicious intent.

Thank you for sharing the response from Mailgun.  Their fix solves the problems very quickly and easily.

**Comment by berksites** [comment]
@michaelphipps I suppose your right - still though, I was on mg 1.7 - instead of using the cert I upgraded to 2.4 and alls well.

**Comment by rickyharerri** [comment]
Simply Goto file "PATH_TO_MAILGUN_FOLDER/mailgun-php/src/Mailgun/Connection/RestClient.php" and change  (Near line number 62)
protected $sslEnabled = true;  
To
protected $sslEnabled = false;

in my case its start working.

**Comment by DavidGarciaCat** [comment]
Hey @rickyharerri 

I think this is not a wise idea. In case you need to run your `composer install|update` commands then you might get an error due new updates on your Mailgun implementation have not been committed, depending on your project's settings.

Given Mailgun team gave an official answer that just works, it's better to follow their suggestion instead of apply new updates to this SDK, due they admitted the updated that they did and the problem that it caused to some implementations.

On the other hand, I didn't get this error when I was running some rests recently, so I must assume that not everybody have this problem and it means we might not need a new PR for this.

Hope this helps.

Cheers,
